ethcore/sync/src/light_sync/warp_sync.rs

```diff
@@ -0,0 +1,652 @@
+
+			trace!(target: "warp", "Requesting a snapshot chunk {:?} from peer {}", &hash, peer_id);
+			let mut rlp = RlpStream::new_list(1);
+			rlp.append(&hash);
+
+			ctx.send(*peer_id, chain::GET_SNAPSHOT_DATA_PACKET, rlp.out());
+		}
+	}
+
+	pub fn disconnect_slowpokes(&mut self, ctx: &SnapshotSyncContext) {
+		let now = Instant::now();
+		let aborting: Vec<PeerId> = self.peers
+			.iter()
+			.filter_map(|(id, peer)| {
+				let elapsed = now - peer.ask_time;
+				let timeout = match peer.asking {
+					None => false,
+					Some(SnapshotPeerAsking::SnapshotData(_)) => elapsed > chain::SNAPSHOT_DATA_TIMEOUT,
+					Some(SnapshotPeerAsking::SnapshotManifest) => elapsed > chain::SNAPSHOT_MANIFEST_TIMEOUT,
+				};
+				match timeout {
+					true => Some(*id),
+					false => None,
+				}
+			})
+			.collect();
+		for peer_id in &aborting {
+			trace!(target: "warp", "Timeout {}", peer_id);
+			ctx.disconnect_peer(*peer_id);
+			self.on_peer_aborting(peer_id);
+		}
+	}
+
+	pub fn timeout(&self) -> bool {
+		self.sync_start_time.map_or(false, |t| t.elapsed() > WAIT_PEERS_TIMEOUT)
+	}
+
+	/// Get the warp sync settings.
+	pub fn warp_sync(&self) -> WarpSync {
+		self.warp_sync
+	}
+
+	/// Clear snapshot state.
+	pub fn clear(&mut self) {
+		self.snapshot.clear();
+	}
+
+	pub fn has_manifest(&self) -> bool {
+		return self.snapshot.have_manifest()
+	}
+
+	pub fn is_complete(&self) -> bool {
+		self.snapshot.is_complete()
+	}
+
+	pub fn done_chunks(&self) -> usize {
+		self.snapshot.done_chunks()
+	}
+
+	/// Validate chunk and mark it as downloaded.
+	pub fn validate_chunk(&mut self, chunk: &[u8]) -> Result<ChunkType, ()> {
+		self.snapshot.validate_chunk(chunk)
+	}
+
+	/// Get the snapshot hash.
+	pub fn snapshot_hash(&self) -> Option<H256> {
+		self.snapshot.snapshot_hash()
+	}
+
+	/// Note snapshot hash as bad.
+	pub fn note_bad(&mut self, hash: H256) {
+		self.snapshot.note_bad(hash)
+	}
+
+	pub fn reset_manifest_to(&mut self, manifest: &ManifestData, hash: &H256) {
+		self.snapshot.reset_to(manifest, hash);
+	}
+
+	pub fn on_peer_status(
+		&mut self,
+		event: &SnapshotSyncEvent,
+		rlp: Rlp,
+		genesis_hash: H256,
+	) -> Result<(), WarpSyncError> {
+		use chain::{PAR_PROTOCOL_VERSION_1, PAR_PROTOCOL_VERSION_2, PAR_PROTOCOL_VERSION_3};
+
+		let supported_versions = [PAR_PROTOCOL_VERSION_1.0, PAR_PROTOCOL_VERSION_2.0, PAR_PROTOCOL_VERSION_3.0];
+		let peer_id = event.peer();
+		let warp_protocol = event.as_context().protocol_version(event.peer()).unwrap_or(0) != 0;
+
+		if !warp_protocol {
+			trace!(target: "warp", "Peer doesn't support warp sync {}", peer_id);
+			return Err(WarpSyncError::NotServer);
+		}
+
+		if self.peers.contains_key(&peer_id) {
+			debug!(target: "warp", "Unexpected status packet from a known peer {}", peer_id);
+			return Ok(());
+		}
+
+		let protocol_version: u8 = rlp.val_at(0)?;
+		let network_id: u64 = rlp.val_at(1)?;
+		let _difficulty: U256 = rlp.val_at(2)?;
+		let _latest_hash: H256 = rlp.val_at(3)?;
```

Comment:
Why fetch these two if we don't need them? (and why bail on error?) Is it faster maybe?

Reply:
To ensure that the status message that we receive from a full node is well formed.